ABM/Modules/Demographic_functions.jl

```diff
@@ -0,0 +1,765 @@
+        b2_jabowas::Float64,
+        b3_jabowas::Float64,
+        max_dbhs::Float64,
+        max_heights::Int64,
+        generic_gradient::Float64
+    )
+        agent.age += 1
+
+        #% CALCULATE COMPETITIVE PENALTY FROM NEIGHBOURS------------#
+        competitive_penalty = 1
+        #* Shade competive penalty applied to trees
+        if shade_height ≥ height
+            shad_comp_val = Float64(abs(Complex(height ./ shade_height) .^ 0.5))
+            competitive_penalty = min(1, (comp_multiplier .* shad_comp_val))
+        end
+
+        #TODO: THIS IS A GENERIC GRADIENT WHICH SHOULD BE REPLACED BY REALISTIC GRADIENTS
+        #* This generic gradient is based on values in each cell drawn from a random uniform
+        #* distribution between 0 and 1. This value is used to in association with an exponential
+        #* distribution to calculate the probability density for that gradient value on that 
+        #* distribution. This probability density is scaled between 0 and 1 and then subtracted
+        #* from 1 to ensure cells with high values (i.e. higher resources) have higher competitive
+        #* values remembering lower values result in lower growth.
+        d = truncated(Exponential(), lower = 0.0, upper = 1.0)
+        g_comp_val = 1 - (pdf(d, generic_gradient) / pdf(d, 0.0))
+
+        #* Select the resource which is most limiting
+        competitive_penalty = min(competitive_penalty, (comp_multiplier .* g_comp_val))
+        #% CALCULATE EDGE PENALTY-----------------------------------#
+        edge_penalty = 1
+        if edge_effects == true
+            edge_penalty = (1 - (edge_weight * (1 - edge_responses)))
+        end
+        
+        #% CALCULATE OVERALL GROWTH PENALTY-------------------------#
+        #* Store growth penalty history for use
+        #* later in suppression mortality
+        #! Note 1 is no change in growth and 0 is no growth
+        growth_reduction = competitive_penalty * edge_penalty
+        prepend!(agent.previous_growth, growth_reduction)
+
+        #* Only the last 5 ticks worth of growth suppression are 
+        #* stored
+        if length(agent.previous_growth) > 5
+            agent.previous_growth = agent.previous_growth[1:5]
+        end
+
+        #% INCREMENT TREE HEIGHT AND DBH----------------------------#
+        # For trees
+        if growth_form == 1
+            #? Equations are based on Botkin et al (1972)
+            dbh_increment = ((dbh * g_jabowas * 
+                    (1 - (dbh * height) / (max_dbhs * max_heights))) /
+                (2.74 + 3 * b2_jabowas * dbh - 4 * b3_jabowas * dbh ^ 2))
+
+            #! Note change of scoping when assigning new values. This is done
+            #! to ensure values are updated in the global pool
+            agent.dbh += (dbh_increment * competitive_penalty * edge_penalty)
+            agent.height = (1.37 + (b2_jabowas * agent.dbh) - 
+                (b3_jabowas * agent.dbh * agent.dbh))
+        
```

Comment:
(**style**) You are mutating a function argument here. See later comment about this.